pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -186,27 +182,32 @@ public GenericRow updateRecord(IndexSegment segment, RecordInfo recordInfo, Gene
       }
     }
 
+    ThreadSafeMutableRoaringBitmap validDocIds = segment.getValidDocIds();
+    assert validDocIds != null;
     _result = record;
-    _primaryKeyToRecordLocationMap
-        .compute(hashPrimaryKey(recordInfo._primaryKey, _hashFunction), (primaryKey, currentRecordLocation) -> {
+    _primaryKeyToRecordLocationMap.compute(hashPrimaryKey(recordInfo._primaryKey, _hashFunction),
+        (primaryKey, currentRecordLocation) -> {
           if (currentRecordLocation != null) {
             // Existing primary key
 
-            // Update the record location when the new comparison value is greater than or equal to the current value
-            // . Update
-            // the record location when there is a tie to keep the newer record.
+            // Update the record location when the new comparison value is greater than or equal to the current value.
+            // Update the record location when there is a tie to keep the newer record.
             if (recordInfo._comparisonValue.compareTo(currentRecordLocation.getComparisonValue()) >= 0) {
               IndexSegment currentSegment = currentRecordLocation.getSegment();
+              int currentDocId = currentRecordLocation.getDocId();
               if (_partialUpsertHandler != null) {
                 // Partial upsert
                 _reuse.clear();
-                GenericRow previousRecord = currentSegment.getRecord(currentRecordLocation.getDocId(), _reuse);
+                GenericRow previousRecord = currentSegment.getRecord(currentDocId, _reuse);
                 _result = _partialUpsertHandler.merge(previousRecord, record);
               }
-              assert currentSegment.getValidDocIds() != null;
```

Comment:
why do we need to move this up? can you add some comment?

Reply:
`assert` is used to suppress the IDE warning on potential NPE without adding overhead. It generally means it is not possible to have `null` validDocIds. I don't think it is worth commenting though

Reply:
I meant `segment.getValidDocIds()` is called earlier

Reply:
Assert statements generally do add some overhead because they add about 10 bytes of bytecode each, which interferes with hotspot’s  bytecode size based inlining heuristics and may result in a method containing asserts not being inlined. I absolutely wouldn’t suggest removing it, but it’s important to know they aren’t free.

Reply:
Also, don’t be concerned about adding null checks, they aren’t compiled the way you see them - they get speculatively optimised away based on profiling data collected in lower tiers. So this is not more efficient than a null check which never fails, it’s exactly as efficient but with more bytecode weight. This is a good read on the topic: https://shipilev.net/jvm/anatomy-quarks/25-implicit-null-checks/

Reply:
@yupeng9 That shouldn't matter as it just reads a member variable from the segment into a local variable. I simply extracted that common logic ahead to avoid making multiple assertions.

@richardstartin Great article, thanks for sharing. In that case should we prefer `Objects.requireNonNull(currentSegment.getValidDocIds()).remove(currentDocId);` as IDE suggested to suppress the warning?

Reply:
@Jackie-Jiang what about just a null check? If it really can't be null then it doesn't change the logic. Also, if it _really_ can't be null, why not annotate the method `@Nonnull`?

Reply:
@richardstartin The method can return `null` for non-upsert case, but in the current context it can never be `null`. `Objects.requireNonNull()` is basically the concise version null check without putting the if/else block